Resource-safe C interface, part 1 (the 4.10 backwards-compatibility edition)

gadmm opened the pull request:
This introduces new functions `caml_process_pending_*` (new public-facing name for `caml_check_urgent_gc`) with variants that return an exception instead of raising it immediately (in the style of `caml_callback_exn`. The overall goal is to allow writing resource-safe C functions in the presence of asynchronous exceptions.

This is a follow-up to #8691 and #8941. #8691 made this proposal possible by limiting the places where exceptions can be raised in C code. ~Unfortunately, it can also break code for users with long-running C functions, who now have to change their code to poll regularly.~ I think this is a sound design, but users need to be accompanied and the new usage has to be documented and given support. Documentation has a PR at #8941 which will need to be updated if this PR is accepted. Support is provided by this PR (part 1), specifically limited to what would be nice to do for 4.10 (in particular providing a transition for the Coq proof assistant). (Note, there are two things: settling on an interface with nice names for the explicit polling, and in addition to that, what is needed to support Coq's use-case. But their current code relies on caml internals, so there is no obligation to be nice.)

This contains a proposal for a public-facing interface. This is best read commit-per-commit, each commit log includes a detailed rationale. Let me know if the github interface makes this impractical.

#### To do

- [x] Documentation (#8941)

gadmm committed:
[minor] typo

gadmm commented:
I think that @jhjourdan and @stedolan can be interested and have opinions on the PR.

An example of proposed use of the new interface is given at https://github.com/coq/coq/issues/10603#issuecomment-536130438.

gadmm commented:
I think @dbuenzli can be interested and have opinions on the public interface.

gadmm committed:
Changes

jhjourdan commented on runtime/caml/signals.h:
```diff
@@ -44,13 +44,11 @@ void caml_record_signal(int signal_number);
 CAMLextern void (*caml_enter_blocking_section_hook)(void);
 CAMLextern void (*caml_leave_blocking_section_hook)(void);
 CAMLextern int (*caml_try_leave_blocking_section_hook)(void);
-CAMLextern void (* volatile caml_async_action_hook)(void);
```
Are we sure no external C code uses this function?

I know this is part of the internal API and we should not worry to much (as for the Coq VM), but perhaps you could search on e.g., Github for other uses, just for making sure.

gadmm replied:
Indeed I did grep my local copy of the opam repository, but I did not mention it because the result were inconclusive. They were inconclusive because I did not find anything, and such a search can only show examples of uses, it cannot let us conclude that something is not used :) The argument is indeed that it's internal.

jhjourdan replied:
Ok, I guess that should be enough checking.

jhjourdan commented on runtime/signals.c:
```diff
@@ -289,50 +300,78 @@ void caml_request_minor_gc (void)
-   from `caml_check_urgent_gc`). */
-void caml_raise_in_async_callback (value exc)
+// backwards-compatibility
+CAMLexport value caml_check_urgent_gc(value extra_root)
```
We probably should dimply use a macro to define the alias instead of this additional function call.

gadmm replied:
The use of macros in the codebase leaves me cold. Maybe using the C99 inline keyword would be better? Is there any guidance on the use of macros in the OCaml codebase?

jhjourdan replied:
I don't think there is any general guidance. I would say that using macros for defining aliases is fine, and this is what `compatibility.h` has done for years for many symbols.

That said, I do not care too strongly, and this may even become impossible if you take into account my request of merging the various functions.

gadmm replied:
Thank you for the review; with the new semantics for check_urgent_gc, the question no longer arises.

jhjourdan commented on runtime/memprof.c:
```diff
@@ -208,14 +210,18 @@ static value do_callback(tag_t tag, uintnat wosize, uintnat occurrences,
   CAMLreturn(res);
 }
 
+static value do_callback(tag_t tag, uintnat wosize, uintnat occurrences,
```
I don't think there is need for this function. Its only use is in `caml_memprof_track_young`, and we will probably need the `exn` variant later for freeing some resources later. So would you please inline it in `caml_memprof_track_young`?

gadmm replied:
done.

jhjourdan commented on runtime/caml/signals.h:
```diff
@@ -30,6 +30,19 @@
+CAMLextern void caml_leave_blocking_section (void);
+
+CAMLextern void caml_process_pending_events (void);
+CAMLextern value caml_process_pending_events_with_root (value extra_root);
```
A note on names: in the OCaml runtime, and "event" usually refers to a "debugging event", which are checkpoints for the bytecode debugger. So we should perhaps use a different name here: "pending callback", for example?

Apart from that, I think I would prefer limiting the number of variants of the same function, for the sake of the ease of navigation of `.h` files. For that purpose, I think we could drop the `_exn` variants and have an additional flag indicating whether or not the exception should be forwarded (with an enum giving names such as `CAML_EXN_RAISE` or `CAML_EXN_FORWARD`). As for the additional root, why not adding an extra `value* extra_root` parameter, which we would set to `NULL` in case we don't want to use this feature? Of course we would need a small compatibility wrapper for `caml_check_urgent_gc`.

gadmm replied:
About "event": ok, a visit to the thesaurus allows me to propose "action" as a substitute.

I am rather unconvinced about the use of flags because:
- they are not going to be used programmatically,
- the flag would need to determine the return type. We could cheat, but this is bad.
On the contrary, I simply reuse the existing convention from `call_callback*_exn`, so the added learning effort for the programmer is minimal.

jhjourdan replied:
The size of the runtime API keeps growing and having several subtly different variants of the same function does not help understanding/remembering it. Sure, naming patterns help, but flags help factorizing the API and the implementation, forces the user to think about the issue of exceptions, and makes the API more extensible. Note that you will probably use them programmatically to forward calls to these functions from one to another.

Actually, I could even propose you to factorize `caml_process_pending_events` and `caml_process_pending_gc` using another flag. The argument is the same : factorize subtly different functions (and thus make it clear that they serve similar purpose), and force the user to think about the value of the flags instead of calling the first functions that seems to fit its need.

gadmm replied:
I have changed the semantics of caml_check_urgent_gc back, and one consequence is that there are a lot fewer variants. I would have liked to avoid the with_root variant and this is now the case given that there are no more uses of `process_pending_events_with_root` in the wild (the existence of such a function was a side-effect of making check_urgent_gc poll). caml_check_urgent_gc is no longer superseded and keeps its original meaning.

In the end, the public interface has only two new functions: `caml_process_pending_actions{,_exn}`. Let me know if this reassures regarding the growing size of API. I plan to keep reusing the same convention in the future (the one from caml_callback{,_exn} and caml_startup{,_exn}).

jhjourdan replied:
Good!

jhjourdan commented on runtime/fail_byt.c:
```diff
@@ -20,9 +20,10 @@
 #include "caml/fail.h"
-#include "caml/io.h"
 #include "caml/gc.h"
+#include "caml/io.h"
```
Is there a rationale for changing the order of imports here?

jhjourdan commented on runtime/fail_nat.c:
```diff
@@ -20,17 +20,17 @@
 #include "caml/fail.h"
-#include "caml/io.h"
 #include "caml/gc.h"
+#include "caml/io.h"
```
Idem.

jhjourdan commented on runtime/fail_nat.c:
```diff
@@ -20,17 +20,17 @@
 #include "caml/stack.h"
-#include "caml/roots.h"
-#include "caml/callback.h"
 
```
Idem.

jhjourdan commented on runtime/fail_nat.c:
```diff
@@ -164,6 +164,11 @@ void caml_raise_sys_blocked_io(void)
+{
+  if (Is_exception_result(res)) caml_raise(Extract_exception(res));
+}
+
```
That's rather unfortunate that we have to copy-paste that in the two files. I know this is not your fault, but we should perhaps have a `fail.c` for the common parts of the two files?

jhjourdan replied:
(Let's do this in a different PR.)

gadmm replied:
ok

jhjourdan commented on runtime/caml/fail.h:
```diff
@@ -61,6 +61,8 @@ struct longjmp_buffer {
 
 int caml_is_special_exception(value exn);
 
+void caml_raise_if_exception (value res);
```
While you are at factorizing, you could make this function return the value being passed as parameter. This would simplify a lot most of the call sites.

gadmm replied:
Good catch, thanks.

jhjourdan commented on runtime/array.c:
```diff
@@ -522,7 +522,7 @@ static value caml_array_gather(intnat num_arrays,
     /* Many caml_initialize in a row can create a lot of old-to-young
        refs.  Give the minor GC a chance to run if it needs to. */
-    res = caml_check_urgent_gc(res);
+    caml_process_pending_events();
```
AFAIU, `res` is not rooted here. Please either make it a root (-> very slight performance degradation in the general case, probably not worth wondering) or pass it as a an extra root.

gadmm replied:
Sorry, my mistake

jhjourdan commented on runtime/intern.c:
```diff
@@ -31,10 +31,12 @@
 #include "caml/mlvalues.h"
 #include "caml/misc.h"
 #include "caml/reverse.h"
-#include "caml/memprof.h"
```
Why changing the order of includes?

gadmm replied:
I understand that the general recommendation is to respect surrounding style of the code, and it appears to me that headers are usually sorted alphabetically, so I am fixing this on the way. (Of course if there is a known bug that the order matters then I will revert those changes, but I would be very surprised.)

jhjourdan replied:
Ah, I never spotted this. I don't really care. You can ignore these comments.

jhjourdan commented on runtime/meta.c:
```diff
@@ -30,9 +32,9 @@
+#include "caml/signals.h"
 #include "caml/stacks.h"
-#include "caml/debugger.h"
-#include "caml/backtrace_prim.h"
```
Idem.

jhjourdan commented on runtime/signals.c:
```diff
@@ -353,13 +353,6 @@ CAMLexport void caml_process_pending_events(void)
   caml_raise_if_exception(caml_process_pending_events_exn());
 }
 
-CAMLexport value caml_process_pending_events_with_root_exn(value extra_root)
```
If this function is finally not useful, please remove it from the commit that introduced it.

gadmm replied:
done

jhjourdan commented on runtime/signals.c:
```diff
@@ -175,7 +174,7 @@ CAMLexport void caml_enter_blocking_section(void)
     /* Check again for pending signals.
        If none, done; otherwise, try again */
-    if (! caml_something_to_do) break;
+    if (!caml_get_urgent_to_do()) break;
```
What is the purpose of this change?

jhjourdan replied:
Sorry for that comment. I misread.

jhjourdan commented on runtime/signals.c:
```diff
@@ -317,6 +322,8 @@ value caml_process_pending_gc_with_root(value extra_root)
 
 CAMLexport value caml_process_pending_events_exn (void)
 {
+  if (!caml_get_urgent_to_do()) return Val_unit;
```
I don't like this change : even thought it may not be bogus in itself, it makes the code way less robust, for two reasons:
- In native mode, if `caml_something_to_do` is not set but the minor GC limit is set to `young_alloc_end`, then we end up in an loop, because we never call `caml_update_young_limit`.
- If there is a pending callback but `caml_something_to_do` is not set, then we need to wait for the next callback to get a chance of executing the first one.

Sure, if we maintain the invariant that `caml_something_to_do` is always set when there is a pending callback, then these two issues can never happen. But this is based on a rather subtle logic, and, moreover, accesses to `caml_something_to_do` and `young_limit` can be racy so that it's not clear that we see the right changes at the right moment.

gadmm replied:
Sorry for the loop, I should have realised because I knew it from my work on implementing masking.

Note that whether the user is tasked in calling caml_get_urgent_to_do themselves or this implementation, the problem is the same. The goal here is to avoid exposing assumptions in the public interface about the race between querying signals and running signal handlers. It is not yet clear to me how multicore is going to affect this, but this way we can change it more easily in the future.

What we should do here is implement it differently in native and bytecode, and in native we check for `young_limit==young_alloc_end` instead (I already have to introduce the symmetric function `caml_urgent_notify_program` for Sys.mask, see https://github.com/ocaml/ocaml/pull/8961/commits/41a8ced2201f98dc32667720ea352ceaed283ab3).

jhjourdan replied:
> What we should do here is implement it differently in native and bytecode, and in native we check for young_limit==young_alloc_end instead (I already have to introduce the symmetric function caml_urgent_notify_program for Sys.mask, see 41a8ced).

I would still prefer not doing this test at all. Again, because of the race, `caml_something_to_do` may not be up-to-date in bytecode mode. The same applies for `young_limit` in native mode. And, moreover, the invariant you are relying on is subtle and may get broken in the future by bogus code. I would prefer being robust.

Do we have an idea of how cost a spurious call to the non-optimized version of this function? How much do we gain from this optimization?

gadmm replied:
For the cost, the argument is that you want a C interface to do like OCaml. For instance you might want to remain reactive in a tight loop. In OCaml, we are happy that polling only costs one comparison in that case. The premiss is that there should be a way to do exactly like OCaml (essentially `if (check()) { perform(); }`).

It is true that checking for `young_limit==young_alloc_end` is not exactly like `young_ptr < young_limit`, but I think making `young_limit==young_alloc_end` an explicit invariant of when pending actions should be performed is technically sound. (I may also be missing your point, in that case could you please give us a more detailed explanation of your objection?)

~Furthermore, in bytecode it is a bit more complicated indeed given that now sometimes the test is `young_ptr < young_limit`, other times it is `caml_something_to_do`, but I would prefer to simplify bytecode by always checking for the same condition (so, `young_limit==young_alloc_end`).~

I can propose to make the non-checking function internal and provide a public function that does the test before calling the internal function.

The following piece of code from `ocamlcc.1.0/runtime/ocamlcc-signals.c` makes me think further (it relies on internals):
```c
  if (caml_something_to_do) {
    long ocamlcc_save_sp_offset;
    SaveSp(ocamlcc_save_sp_offset);
    sp[-frame_sz - 1] = sp[-frame_sz - 2] = sp[-frame_sz - 3] =
      sp[-frame_sz - 4] = sp[-frame_sz - 5] =
      sp[-frame_sz - 6] = Val_unit;
    OffsetSp(-frame_sz - 6);
    do {
      caml_something_to_do = 0;
      caml_process_event();
    } while (caml_something_to_do);
    ResetSp(-frame_sz - 6);
    RestoreSp(ocamlcc_save_sp_offset);
  }
```
If we do want to support the inner loop, it looks to me that we need to expose `caml_get_urgent_to_do()`, contrary to what I advocate. But I do not know if we want to support something like it. We can keep `caml_get_urgent_to_do` internal, since they already rely on internals.

We will see what I come up with.

gadmm replied:
I initially wrote that in bytecode sometimes the test is `young_ptr < young_limit`, but this is currently never because `young_limit == young_alloc_end`. Similarly in native code we can get to caml_garbage_collection by exceeding `caml_memprof_young_trigger`. In that case, as I now understand, it is not only brittle but wrong to check `young_limit == young_alloc_end`.

jhjourdan replied:
My point is that, in rare situations, `caml_something_to_do` and `young_limit` may not be set even though there is an async callback waiting. I can foresee several reasons:
- Because of the race, writes to these variables may not be visible to the OCaml thread at the right moment. Imagine we get into `caml_check_urgent_gc` when `caml_something_to_do` is seen as set but not `caml_pending_signals`: we may miss a signal in the first call to `caml_check_urgent_gc` and don't check for them later because we return from said function even before checking `caml_pending_signals`.
- The fact that `caml_something_to_do` is set when there are pending callbacks relies on subtle invariants which we will easily break in the future.
- If there are two different callbacks (say, a signal and a finalizer) arriving at the same time, then we set `caml_something_to_do` to 0 when starting the first one while the second one is still waiting. We may want to run the second one if the first one is taking  long.

In the current situation, the code is robust in the sense that async callbacks are called at the next minor collection even if, for some reason, `caml_something_to_do` is not set properly. With what you are proposing, we have to wait for the next async callback, which may not come at all.

> I can propose to make the non-checking function internal and provide a public function that does the test before calling the internal function.

Perhaps what we should do then is make `caml_something_to_do` part of the public interface and document that this is the thing one has to check before calling `caml_check_urgent_gc`. This would both cover ocamlcc and Coq use cases, and make polling very cheap as you desire.

Note, however, that because of the reasons I described above, it may be possible that we miss async callbacks in long-running C code. I guess that's the price to pay?

gadmm replied:
You explain well why we need to distinguish the function called internally from the function provided to the user (which will be guarded by a check to caml_something_to_do).

With masking, we will need to make it a hard invariant that caml_something_to_do has to be set if there are pending callbacks (outside the case of process_pending* where it is temporarily set to zero and set back to 1 later if needed). I think it is better to start doing it now.

In the case of a finaliser taking a long time while a signal is pending and caml_something_to_do is temporarily 0 (or vice-versa), we can think further about it but I think this is outside the scope.

These reasons convince me to _not_ make public a check for caml_something_to_do, so that we keep the races under our control and it is easier to fix any issue later.

I am not sure we want to support this exact code of ocamlcc, because I don't think there is a meaning to "please run signals until there are no more".

I am not sure I see under which conditions async callbacks could be forgotten, what scenario do you think about exactly?

jhjourdan replied:
> With masking, we will need to make it a hard invariant that caml_something_to_do has to be set if there are pending callbacks (outside the case of process_pending* where it is temporarily set to zero and set back to 1 later if needed). I think it is better to start doing it now.

Why so? Cannot we simply call `caml_check_urgent_gc` after any unmasking?

> These reasons convince me to not make public a check for caml_something_to_do, so that we keep the races under our control and it is easier to fix any issue later.

What about exposing a getter then? This would make it easy to change its content if needed. This getter could even be an inline function, for performance.

> I am not sure we want to support this exact code of ocamlcc, because I don't think there is a meaning to "please run signals until there are no more".

That's not what worries me in ocamlcc's use case. What worries me is that it seems important for its performance not to register its extra roots and perform the function call at every single check for signals.

> I am not sure I see under which conditions async callbacks could be forgotten, what scenario do you think about exactly?

I don't get your question: I thought my previous message was all about answering it... If `caml_something_to_do` is 0 when there is a pending async callback, then, with your proposal, we would need to check for the next async callback to come for executing the first one. To me, this is essentially as if we lost it.

jhjourdan replied:
Just to make things clear: my current proposal is to expose in the public interface a getter for `caml_something_to_do`, and document that it can be used for fast polling in C code. This getter could be an inline function in the `signals.h` file to make its cost minimal. We would *not* check for `caml_something_to_do` at every minor GC, just to make sure we don't miss async callbacks.

Do you agree with this proposal? If not, then, what is the issue? It seems to me this addresses all your concerns.

gadmm replied:
Ok, I thought you meant that some signal would be dropped. Now I understand the issue with long-running C code, this is well spotted. I am proposing the following fix: allow users the obtain the same fallback behaviour as there is from OCaml if they cooperate, by setting caml_something_to_do at every minor collection or major slice in CAML_FROM_C mode.

Waiting for a minor collection or a major slice is still bad latency. It would be nice to fix the first race you explain between caml_something_to_do and caml_pending_signals (this beyond my expertise). I imagine that something will have to be done anyway with multicore (I thought I saw it mentioned at some point but could not find a dedicated bug report). (I am also thinking that caml_signals_are_pending might be redundant now with caml_something_to_do, but I am not sure I should do something about it right now.)

I am wary of exposing a getter for caml_something_to_do as much as exposing caml_something_to_do itself: I want to avoid hardcoding backwards-compatibility support for the form `if (check()) { execute(); }` and everything else users can come up with. This would mean losing control over the critical section between `check` and `execute`, limiting future evolution. In particular this limits the space of possible solutions of the two problems and of any other that could arise (allow me to remain abstract; if I had a concrete solution in mind right now I would propose it). Anyway, I do not see a need for it compared to your proposal now that we can replicate the fallback behaviour in caml_process_pending_actions.

Ocamlcc is a special because it is an OCaml compiler so it is not excessive to ask that they keep relying on internals. It is like saying that the public interface does not support the use-case from `runtime/interp.c` either. The goal of this new API is not to support the creation of alternative OCaml compilers, as this could require to make public many more internal runtime identifiers beyond those.

I hope the new version of the code addresses all your related concerns (I am waiting for tests to finish). I am documenting the invariant for caml_something_to_do so that we go towards fixing it and not breaking it.

(About unmasking: we do not execute callbacks right away for crucial reasons related to reasoning about critical sections for polling. I gave a thought to what made OCaml's polling behaviour nice to reason about in my experience, and I wrote it down here: https://github.com/ocaml-multicore/ocaml-multicore/issues/187#issuecomment-535148688. It is entirely about almost your question. But I guess you just want to poll for signals unconditionally when unmasking and not rely on caml_something_to_do to the extent that it is unreliable: that is something else and very reasonable, do not hesitate to bring up this point again at #8961.)

jhjourdan commented on Changes:
```diff
@@ -15,6 +15,14 @@ Working version
+  any exception arising during their execution, so that resources can
+  be cleaned-up before re-raising it.
+  (Guillaume Munch-Maccagnoni, review by TODO)
+
```
You also need to fix the doc you wrote for #8941, since `caml_check_urgent_gc` is now deprecated.

gadmm replied:
Done

jhjourdan reviewed (changes requested):
I think this PR goes to the right direction in exposing the `_exn` variants of these functions.

I left a few comments in the diff. I particular, I would prefer not introducing several variants for functions that essentially do the same thing. I instead prefer using flags or optional parameters.

jhjourdan commented:
And, BTW, thanks for this PR, this is quite a lot of well-polished work.

gadmm commented on runtime/signals.c:
```diff
@@ -280,59 +290,87 @@ void caml_update_young_limit (void)
+  res = caml_final_do_calls_exn();
+  if (Is_exception_result(res)) goto exception;
+
+  // Call exception handlers
```
Note to myself: this should be "call signal handlers"

gadmm commented:
I'll get back to this shortly. In the meanwhile, I note that some comments are about choices for a public-facing interface. I would like to mention that anybody should feel free to contribute to these, not just the experts of this code.

gadmm commented:
Reminder for myself again: we agreed to change the default behavior of `caml_check_urgent_gc` over at #8691. It has to be done for 4.10 so I will have to split it in case the current PR is not ready for 4.10. (ETA: during this week)

jhjourdan commented:
Do you need help to get this merged this week?

gadmm commented:
No problem to finish it this week, but it would help to have a clear idea of the deadlines in general. Right now it feels like everybody around me is in a hurry and I don't know why. I asked privately about this and I was told "feature freeze somewhere around mid-October", which makes me think we still have time to finish it this week and discuss it comfortably, but it's rather vague.

stedolan commented:
This PR looks nice, but it adds quite a few more functions to the C API. In particular, what's the motivation for providing both `caml_process_pending_gc` and `caml_process_pending_gc_with_root`?

If it's purely to avoid the overhead of adding an extra root, then I'd like to suggest the following implementation of `caml_process_pending_gc(value)` instead:
```c
value caml_process_pending_gc(value extra_root)
{
  if (Caml_state->requested_major_slice || Caml_state->requested_minor_gc) {
    CAMLparam1 (extra_root);
    CAML_INSTR_INT ("force_minor/caml_process_pending_gc@", 1);
    caml_gc_dispatch();
    CAMLdrop;
  }
  return extra_root;
}
```
This version incurs the small overhead of adding a root only when actually invoking the GC, which is an expensive enough operation that the overhead doesn't matter.

gadmm commented:
The motivation was backwards compatibility: the with_root version was an alias for caml_check_urgent_gc. The one I would like to expose does not take an additional root (and the reason it's enough is close to your suggestion that additional roots are easy to add locally). Anyway, the decision to change the semantics of caml_check_urgent_gc will bring a few changes and simplifications. I'll keep your tip in mind to use if need be.

gadmm committed:
[minor] When calling gc_dispatch directly, avoid setting caml_something_to_do

caml_minor_collection is not a new function, it is already used in the
wild and used to be exposed in caml/minor_gc.h in the past. The change
simply amounts to replacing a call to caml_request_minor_gc() with an
assignment requested_minor_gc = 1, bypassing
caml_set_something_to_do().

gadmm committed:
In long-running C code, force examining all callbacks at the next safe
point following every minor collection or major slice.

Also run signal handlers first.

Indeed, in some cases, caml_something_to_do is not reliable (spotted
by @jhjourdan):

* We could get into caml_process_pending_actions when
  caml_something_to_do is seen as set but not caml_pending_signals,
  making us miss the signal.

* If there are two different callbacks (say, a signal and a finaliser)
  arriving at the same time, then we set caml_something_to_do to 0
  when starting the first one while the second one is still waiting.
  We may want to run the second one if the first one is taking long.

In the latter case, the additional fix is to favour signals, which
have a lower latency requirement, whereas the latency of finalisers
keeps the same order of magnitude, and memprof callbacks are served on
a best-effort basis.

gadmm commented:
Here is the new version. I will open a different PR for the commit that changes `caml_check_urgent_gc`, because it is a separate entry in the Changes file. That current PR is based on top of that other one.

I have to admit that I do not understand the CAMLno_tsan business, thanks if you could have a look and explain any possible mistake.

gadmm commented:
The first 4 commits are proposed at https://github.com/ocaml/ocaml/pull/9037. This PR and review concerns the 5 last commits.

jhjourdan commented:
> I have to admit that I do not understand the CAMLno_tsan business, thanks if you could have a look and explain any possible mistake.

Don't worry too much about that, CI will tell you if you do it wrong.

jhjourdan commented on Changes:
```diff
@@ -18,6 +18,16 @@ Working version
 - #6662, #8908: Add "module _ = X" syntax
   (Thomas Refis, review by Gabriel Radanne)
 
+- #8993: New C functions caml_process_pending_actions{,_exn} in
```
Small nitpick: I thought `caml_process_pending_actions` were introduced in #9037. BTW, I don't really see the rationale behind using two PR here, but I don't care.

jhjourdan commented on runtime/memprof.c:
```diff
@@ -141,9 +141,10 @@ CAMLprim value caml_memprof_set(value v)
+  if (!caml_memprof_suspended) {
+    value exn = caml_memprof_handle_postponed_exn();
+    if (Is_exception_result(exn)) caml_raise(Extract_exception(exn));
+  } else
```
Cannot you use `caml_raise_if_exception` here instead?

gadmm replied:
thanks for spotting it

jhjourdan commented on runtime/memprof.c:
```diff
@@ -182,14 +183,14 @@ enum ml_alloc_kind {
-static value do_callback(tag_t tag, uintnat wosize, uintnat occurrences,
-                         value callstack, enum ml_alloc_kind cb_kind) {
+/* When we call do_callback_exn, we suspend/resume sampling. In order
+   to to avoid a systematic unnecessary polling after each memprof
```
Typo (my bad) : "to to".

jhjourdan commented on runtime/caml/signals.h:
```diff
@@ -45,6 +45,27 @@ CAMLextern value caml_process_pending_actions_exn (void);
+      finaliser) arriving at the same time, then we set
+      caml_something_to_do to 0 when starting the first one while the
+      second one is still waiting. We may want to run the second one
+      if the first one is taking long.
```
Not sure this deserves a FIXME. The current solution is rather good, and I don't think there is anything better we could do.

gadmm replied:
IIUC, in these situations, instead of reacting at the next allocation, it react on the order of 10^5 to 10^6 allocations later. For finalisers the total latency remains on the same order, but for signals this is bad. In particular, have a look at the part of https://github.com/ocaml-multicore/ocaml-multicore/issues/187#issuecomment-532521642 which says "guaranteed interrupt at fixed delays... Based on the maximum latency of interrupt required...". From their point of view of guaranteed latency, this is still buggy.

The second bullet point can be solved if we call signal handlers first, what do you think about it?

It would be nice to fix the first one as well.

jhjourdan replied:
> The second bullet point can be solved if we call signal handlers first, what do you think about it?

Alright, let's do that. Note, however, that the issue is still there for finalisers or memprof. That's sort of an unsolvable problem.

Anyway, for both issues, they will only occur at very rare occasions, and we have a workaround. And it's unlikely we will find a "fix" any time. Hence I don't think the "FIXME" is deserved.

But that's a detail, I don't care, let's move forward.

gadmm replied:
The second bullet point has been fixed. Signal handlers are now safe from this (newly-introduced, AFAIU) race. It is no longer a FIXME, because it is not a bug for finalisers (their latency remains on the same order of magnitude), and it is not a bug for memprof callbacks either, because although we want them to run ASAP, they are served on a best-effort basis.

I keep the first item as a FIXME because it is a bug if we take into account code that tries to allocate little, or if there are latency guarantees as assumed in the work on better safe points. If it is a deviation to the intended behaviour and it does not appear easily solvable with the current mechanism, then I think a FIXME is the best option.

jhjourdan reviewed (approved):
The part of this PR which is not also in #9037 looks good. I have minor comments. Nothing important. When #9037 will be merged, this one is ready to merge.

jhjourdan commented:
Windows CI is failing, it is probably due to a segfault, but I am not sure. Let's see what happens when you will update this.

gadmm committed:
Resource-safe C interface for async callbacks

Introduce caml_process_pending_actions and
caml_process_pending_actions_exn: a variant of the former which does
not raise but returns a value that has to be checked against
Is_exception_value.

I keep the current conventions from caml_callback{,_exn}: For a
resource-safe interface, we mostly care about the _exn variants, but
every time there is a public _exn function I provide a function that
raises directly for convenience.

They are introduced and documented in caml/signals.h.

Private functions are converted to their _exn variant on the way as
needed: for internal functions of the runtime, it is desirable to go
towards a complete elimination of functions that raise implicitly.

Get rid of the distant logic of caml_raise_in_async_callback. Instead,
caml_process_pending_events takes care itself of its something_to_do
"resource". This avoids calling the former function in places
unrelated to asynchronous callbacks.

gadmm committed:
Move exception encoding helpers from caml/callback.h to caml/mlvalues.h

gadmm commented:
Is it also possible that the Windows CI failing is unrelated to the PR? Otherwise it is probably not fixed in this new version.

Users need to access `Is_exception_result` and `Extract_exception` from signals.h. I moved their definition from callback.h to mlvalues.h. In case this introduction of non-namespaced macros in the general scope is seen as breaking for users (is it?), a milder version would be to include callback.h in signals.h.

jhjourdan commented:
That's a different test, so I would say it *is* related to this PR. I have no idea how, though!

gadmm commented:
I meant the previous one that looked like a segfault. I'll have a look at this one.

jhjourdan commented:
I think I found the issue, it's in the way you are polling in `weak.c`. Let me propose a patch.

jhjourdan commented:
The issue was catched by valgrind. It is fixed by the following patch:
```
diff --git a/runtime/weak.c b/runtime/weak.c
index 2a611b604..e9234bdfc 100644
--- a/runtime/weak.c
+++ b/runtime/weak.c
@@ -292,6 +292,9 @@ static value optionalize(int status, value *x)
   } else {
     res = None_val;
   }
+  // run memprof callbacks both for the option we are allocating here
+  // and the calling function.
+  caml_check_urgent_gc_and_callbacks(Val_unit);
   CAMLreturn(res);
 }
 
@@ -421,8 +424,6 @@ CAMLprim value caml_ephe_get_key_copy (value ar, value n)
 {
   value key;
   int status = caml_ephemeron_get_key_copy(ar, Long_val(n), &key);
-  // run memprof callbacks
-  key = caml_check_urgent_gc_and_callbacks(key);
   return optionalize(status, &key);
 }
 
@@ -478,8 +479,6 @@ CAMLprim value caml_ephe_get_data_copy (value ar)
 {
   value data;
   int status = caml_ephemeron_get_data_copy(ar, &data);
-  // run memprof callbacks
-  data = caml_check_urgent_gc_and_callbacks(data);
   return optionalize(status, &data);
 }
```

jhjourdan commented:
The issue was that `data` and `key` were not necessarily initialized when calling `caml_check_urgent_gc_and_callbacks` on it.

gadmm commented:
Thank you for the bugfix (FTR it is now applied at #9037).

I have rebased. This should be good to go, after the patch that makes `Is_exception_result`/`Extract_exception available` to users of signals.h is reviewed. If there are backwards-compatibility concerns, I thought of a second alternative which is instead to introduce namespaced inline functions in mlvalues.h (`caml_is_exception_result`/`caml_extract_exception`). If so I could update the docs in a second time.

gadmm commented:
#9037 has been merged, so this is now based on trunk.

jhjourdan commented on runtime/signals.c:
```diff
@@ -287,43 +290,68 @@ void caml_request_minor_gc (void)
+{
+  value exn = caml_check_urgent_gc_and_callbacks_exn(Val_unit);
+  caml_raise_if_exception(exn);
 }
```
I'm sorry, but I don't get the point of having these two functions, which are (appart from the extra root), aliases of `caml_check_urgent_gc_and_callbacks{_exn}`. Could you choose an API (either with an extra root or without), and have only one version of these functions?

gadmm replied:
For deciding on a public interface, I adopt the point of view of the user. What they see is that `caml/signals.h` grows from:
```c
CAMLextern void caml_enter_blocking_section (void);
CAMLextern void caml_leave_blocking_section (void);
```
to:
```c
CAMLextern void caml_enter_blocking_section (void);
CAMLextern void caml_leave_blocking_section (void);

CAMLextern void caml_process_pending_actions (void);
/* Checks for pending actions and executes them. This includes pending
   minor and major collections, signal handlers, finalisers, and
   Memprof callbacks. Assumes that the runtime lock is held. Can raise
   exceptions asynchronously into OCaml code. */

CAMLextern value caml_process_pending_actions_exn (void);
/* Same as [caml_process_pending_actions], but returns the exception
   if any (otherwise returns [Val_unit]). */
```
They do not care that for internal and historical reasons we ended up with `caml_check_urgent_gc_and_callbacks` as it is today. Besides the name which I choose to be a bit better since it is a blank slate (from the user POV), I think there is another important difference there: `caml_check_urgent_gc_and_callbacks` takes an extra root because in many of the primitives where it is used, the values are not rooted and we need only one root momentarily. Maybe this was some form of optimisation. But this is not what is advocated by the FFI documentation, and if the user follows the manual (Section 20.6) then they will never need an extra root like that.

Once we've decided for a public API that does not take an extra root, we can ask, can we simplify the internals using the public API, and remove `caml_check_urgent_gc_and_callbacks`? This is a natural question, and it is debatable. This concerns a few `array.c` primitives, `ephemeron_create` in `weak.c`, and `intern_end` in `intern.c` where I have to add roots manually and unconditionally. Remember, I do not know if there are performance implications; by default I assume that they are written in this special style for a reason. So instead, I used Stephen's trick to avoid rooting the value when not needed.

Since you agree with the technical soundness, to move forward, we now need to agree on the public interface. I propose it as above. Anybody interested can chime in.

Once we agree on the public interface, I am interested in hearing your nitpicks (for instance I can remove `caml_check_urgent_gc_and_callbacks{,_exn}` if you tell me that there are no performance implications).

jhjourdan reviewed (commented):
I made the review: Overall, it looks good (I already reviewed it, so...), except I don't understand the point of defining `caml_process_pending_actions{_exn}` as aliases of existing functions.

gadmm commented:
I would like to hear whether moving `Is_exception_result`/`Make_exception_result`/`Extract_exception` from `caml/callback.h` to `caml/mlvalues.h` deserves a “`*`” in the Changes file due to the introduction of non-namespaced macros in the general scope for all users of the FFI.

gadmm commented:
I proofread the whole patch with a fresh mind. I added a missing CAMLreturn (and checked that there are no similar issue anywhere else).

jhjourdan commented:
Alright, let me do a few benchmarks.

jhjourdan commented:
I am OK with removing the functions with the extra root. Here is the patch I propose:

```
diff --git a/runtime/array.c b/runtime/array.c
index f84e365fc..bf509e36f 100644
--- a/runtime/array.c
+++ b/runtime/array.c
@@ -279,8 +279,17 @@ CAMLprim value caml_floatarray_create(value len)
   else {
     result = caml_alloc_shr (wosize, Double_array_tag);
   }
+
   // Give the GC a chance to run, and run memprof callbacks
-  return caml_check_urgent_gc_and_callbacks (result);
+  // This is an inlined version of [caml_process_pending_actions], to
+  // avoid unconditionally registering result as a root.
+  if(caml_something_to_do) {
+    CAMLparam1(result);
+    caml_raise_if_exception(caml_do_urgent_gc_and_callbacks_exn());
+    CAMLdrop;
+  }
+
+  return result;
 }
 
 /* [len] is a [value] representing number of words or floats */
@@ -331,7 +340,8 @@ CAMLprim value caml_make_vect(value len, value init)
     }
   }
   // Give the GC a chance to run, and run memprof callbacks
-  CAMLreturn (caml_check_urgent_gc_and_callbacks(res));
+  caml_process_pending_actions();
+  CAMLreturn (res);
 }
 
 /* [len] is a [value] representing number of floats */
@@ -384,7 +394,8 @@ CAMLprim value caml_make_array(value init)
         Store_double_flat_field(res, i, d);
       }
       // run memprof callbacks
-      CAMLreturn (caml_check_urgent_gc_and_callbacks(res));
+      caml_process_pending_actions();
+      CAMLreturn (res);
     }
   }
 #else
@@ -443,7 +454,7 @@ CAMLprim value caml_array_blit(value a1, value ofs1, value a2, value ofs2,
   }
   /* Many caml_modify in a row can create a lot of old-to-young refs.
      Give the minor GC a chance to run if it needs to. */
-  caml_check_urgent_gc(Val_unit);
+  caml_process_pending_actions();
   return Val_unit;
 }
 
@@ -522,7 +533,7 @@ static value caml_array_gather(intnat num_arrays,
     /* Many caml_initialize in a row can create a lot of old-to-young
        refs.  Give the minor GC a chance to run if it needs to.
        Run memprof callback for the major allocation. */
-    res = caml_check_urgent_gc_and_callbacks (res);
+    caml_process_pending_actions();
   }
   CAMLreturn (res);
 }
diff --git a/runtime/caml/signals.h b/runtime/caml/signals.h
index c2eb9f7ce..3e7872ef3 100644
--- a/runtime/caml/signals.h
+++ b/runtime/caml/signals.h
@@ -76,7 +76,6 @@ void caml_record_signal(int signal_number);
 value caml_process_pending_signals_exn(void);
 void caml_set_action_pending (void);
 value caml_do_urgent_gc_and_callbacks_exn (void);
-value caml_check_urgent_gc_and_callbacks (value extra_root); // raises
 int caml_set_signal_action(int signo, int action);
 void caml_setup_stack_overflow_detection(void);
 
diff --git a/runtime/intern.c b/runtime/intern.c
index e49e20d32..0637b27db 100644
--- a/runtime/intern.c
+++ b/runtime/intern.c
@@ -693,6 +693,7 @@ static header_t* intern_add_to_heap(mlsize_t whsize)
 }
 
 static value intern_end(value res, mlsize_t whsize) {
+  CAMLparam1(res);
   header_t *block = intern_add_to_heap(whsize);
   header_t *blockend = intern_dest;
 
@@ -705,7 +706,8 @@ static value intern_end(value res, mlsize_t whsize) {
     caml_memprof_track_interned(block, blockend);
 
   // Give gc a chance to run, and run memprof callbacks
-  return caml_check_urgent_gc_and_callbacks(res);
+  caml_process_pending_actions();
+  CAMLreturn(res);
 }
 
 /* Parsing the header */
diff --git a/runtime/interp.c b/runtime/interp.c
index b3cea3d6b..ac549b455 100644
--- a/runtime/interp.c
+++ b/runtime/interp.c
@@ -913,7 +913,7 @@ value caml_interprete(code_t prog, asize_t prog_size)
 
     process_actions:
       Setup_for_event;
-      caml_check_urgent_gc_and_callbacks(Val_unit);
+      caml_process_pending_actions();
       Restore_after_event;
       Next;
 
diff --git a/runtime/meta.c b/runtime/meta.c
index d536b68bc..282833287 100644
--- a/runtime/meta.c
+++ b/runtime/meta.c
@@ -186,7 +186,8 @@ CAMLprim value caml_realloc_global(value size)
       Field (new_global_data, i) = Val_long (0);
     }
     // Give gc a chance to run, and run memprof callbacks
-    caml_global_data = caml_check_urgent_gc_and_callbacks(new_global_data);
+    caml_global_data = new_global_data;
+    caml_process_pending_actions();
   }
   return Val_unit;
 }
diff --git a/runtime/obj.c b/runtime/obj.c
index 5c77c2893..d73595dc0 100644
--- a/runtime/obj.c
+++ b/runtime/obj.c
@@ -120,7 +120,7 @@ CAMLprim value caml_obj_with_tag(value new_tag_v, value arg)
     res = caml_alloc_shr(sz, tg);
     for (i = 0; i < sz; i++) caml_initialize(&Field(res, i), Field(arg, i));
     // Give gc a chance to run, and run memprof callbacks
-    res = caml_check_urgent_gc_and_callbacks(res);
+    caml_process_pending_actions();
   }
   CAMLreturn (res);
 }
diff --git a/runtime/signals.c b/runtime/signals.c
index 8f78dfc77..04c18a1db 100644
--- a/runtime/signals.c
+++ b/runtime/signals.c
@@ -122,7 +122,7 @@ void caml_set_action_pending(void)
 /* Record the delivery of a signal, and arrange for it to be processed
    as soon as possible:
    - via caml_something_to_do, processed in
-     caml_check_urgent_gc_and_callbacks_exn.
+     caml_process_pending_actions_exn.
    - by playing with the allocation limit, processed in
      caml_garbage_collection and caml_alloc_small_dispatch.
 */
@@ -324,34 +324,16 @@ exception:
   return exn;
 }
 
-CAMLno_tsan /* The access to [caml_something_to_do] is not synchronized. */
-static inline value caml_check_urgent_gc_and_callbacks_exn(value extra_root)
+CAMLexport inline value caml_process_pending_actions_exn(void)
 {
-  if (caml_something_to_do) {
-    CAMLparam1(extra_root);
-    value exn = caml_do_urgent_gc_and_callbacks_exn();
-    if (Is_exception_result(exn))
-      CAMLreturn(exn);
-    CAMLdrop;
-  }
-  return extra_root;
-}
-
-value caml_check_urgent_gc_and_callbacks(value extra_root)
-{
-  value res = caml_check_urgent_gc_and_callbacks_exn(extra_root);
-  return caml_raise_if_exception(res);
-}
-
-CAMLexport value caml_process_pending_actions_exn(void)
-{
-  return caml_check_urgent_gc_and_callbacks_exn(Val_unit);
+  if (caml_something_to_do)
+    return caml_do_urgent_gc_and_callbacks_exn();
+  return Val_unit;
 }
 
 CAMLexport void caml_process_pending_actions(void)
 {
-  value exn = caml_check_urgent_gc_and_callbacks_exn(Val_unit);
-  caml_raise_if_exception(exn);
+  caml_raise_if_exception(caml_process_pending_actions_exn());
 }
 
 /* OS-independent numbering of signals */
diff --git a/runtime/weak.c b/runtime/weak.c
index e9234bdfc..ee5094cc5 100644
--- a/runtime/weak.c
+++ b/runtime/weak.c
@@ -110,9 +110,12 @@ CAMLexport value caml_ephemeron_create (mlsize_t len)
 
 CAMLprim value caml_ephe_create (value len)
 {
-  value res = caml_ephemeron_create(Long_val(len));
+  CAMLparam0();
+  CAMLlocal1(res);
+  res = caml_ephemeron_create(Long_val(len));
   // run memprof callbacks
-  return caml_check_urgent_gc_and_callbacks(res);
+  caml_process_pending_actions();
+  CAMLreturn(res);
 }
 
 CAMLprim value caml_weak_create (value len)
@@ -294,7 +297,7 @@ static value optionalize(int status, value *x)
   }
   // run memprof callbacks both for the option we are allocating here
   // and the calling function.
-  caml_check_urgent_gc_and_callbacks(Val_unit);
+  caml_process_pending_actions();
   CAMLreturn(res);
 }
```

I had to add roots at three places:
- in `intern_end` and `caml_ephe_create`: since the overhead of these functions is already quite large, I doubt adding a root can have any significant performacne change
- in `caml_floatarray_create`: I am not so sure about whether or not this can really have a performance impact on an actual program. Just in case, I added a manual check to `caml_something_to_do` before registering the root. The performance degradation is about 5% on the following unrealistic micro benchmark:

```ocaml
let () =
  for i = 0 to 100000000 do
    ignore (Float.Array.create 2)
  done
```

I think this is acceptable. Nobody will have such a tight allocation pattern.

Do you agree with this change?

gadmm commented:
Ok, I worked on a patch in the meanwhile. Let us cross-check our patches.

gadmm commented:
I'm OK with the change, but for `caml_floatarray_create`, is the performance degradation with or without the conditional rooting? I agree that nobody will have code like that, so I would prefer to avoid inlining.

gadmm commented:
Is the change to `caml_array_blit` intentional?

gasche commented:
Re `caml_float_array_create`, on the nay-sayer side:
- I could think of some (strange but realistic) programming patterns that involves creating lots of small float arrays (eg. representing 2D points as 2-sized arrays, maybe as the natural result of instantiating a dimension-generic library) with only a couple float operations between them.
- This benchmark is done with the current implementation of the root-registration API. I can't tell if other GC designs could make it costlier (my understanding is that Multicore reuses the same minor heaps, but maybe there is some extra thread-local-access cost?); we have also discussed safer root-manipulation APIs (eg. [camlroot](https://arxiv.org/abs/1812.04905)) which would have a different cost profile, for example in a specific debug mode.

The new API proposed by Guillaume looks nice, and I'm not saying this to discourage you from simplifying things -- a 5% degradation on a microbenchmark does sound reasonable in favor of better ressource-safety in the runtime. But optimizing the runtime code also makes sense.

jhjourdan commented:
> Is the change to caml_array_blit intentional?

No, that's a mistake.

jhjourdan commented:
> I'm OK with the change, but for caml_floatarray_create, is the performance degradation with or without the conditional rooting?

It's with the conditional rooting. I am not sure whether it is noise (the stack frame is different, this can have impact on alignment), or whether there is another effect (the size of the code is larger -> impact on code cache, the stack frame is larger...).

jhjourdan commented:
In any case, I don't think the performance degradation of either inilining or not inlining is worth carrying. So do as you prefer.

But I don't really like the idea of keeping the two aliases just for one function used in one very particular way. This is not mentionning the fact that virtually nobody use `floatarray` directly (people use `float array`, which are created using Array.* instead).

gadmm commented:
> It's with the conditional rooting. I am not sure whether it is noise (the stack frame is different, this can have impact on alignment), or whether there is another effect (the size of the code is larger -> impact on code cache, the stack frame is larger...).

I can believe this, and many other effects. There was an interesting talk at Strange Loop about this (https://www.youtube.com/watch?v=r-TLSBdHe1A, I especially like the part where a difference in benchmarks was due to them running in sessions with user names with different lengths).

BTW it has nothing to do with the absence of CAMLno_tsan? That does not affect generated code?

IIUC you also need to add a root to `caml_array_gather`, used via `Array.sub`, `Array.append`, and `Array.concat`. I am not sure either if the impact is worth caring, but alternatively with two occurrences it justifies a dedicated function. BTW I do not like it either because I would like to remove all implicitly-raising internal functions, so if you tell me that the impact for those three functions is negligible as well _without inlining by hand_ then I prefer to remove it. In the meanwhile, and also because perhaps there is also some convenience value to this function, here is a new version I am satisfied with.

gadmm commented:
> This benchmark is done with the current implementation of the root-registration API. I can't tell if other GC designs could make it costlier

According to @stedolan, there is no reason that making a GC root should be costly, if somebody wanted to work on it.

jhjourdan commented:
> BTW it has nothing to do with the absence of CAMLno_tsan? That does not affect generated code?

This only affect the generated code when the thread sanitizer is activated. Well, at least, I would consider it to be a bug...

> IIUC you also need to add a root to `caml_array_gather`, used via `Array.sub`, `Array.append`, and `Array.concat`.

Indeed, well spotted!

> BTW I do not like it either because I would like to remove all implicitly-raising internal functions, so if you tell me that the impact for those three functions is negligible as well without inlining by hand then I prefer to remove it.

For `caml_array_gather`, the infrastructure around it is much heavier, and the function already registers some roots, so that the overhead is really negligible

About `caml_floatarray_create`, if we don't inline the function, the impact on said micro benchmark is closer to 20%. Given that this function has been specifically optimized (cf #6180) , I would prefer not doing this.

So, I see two solutions:
- Either do the inlining as I proposed above
- Keep the function with the extra root as you propose, but then please give it a name which make it clear that it is very close to the other (e.g., `with_root` suffix), and define it from the `process_actions` functions instead of the other way around, to avoid uselessly registering a root containing `Val_unit` if not needed.

gadmm commented:
> About `caml_floatarray_create`, if we don't inline the function, the impact on said micro benchmark is closer to 20%. Given that this function has been specifically optimized (cf #6180) , I would prefer not doing this.

Ah good to know, thanks for digging up a reference.

> Keep the function with the extra root as you propose, but then please give it a name which make it clear that it is very close to the other (e.g., `with_root` suffix),

Yes, see the latest version, this is what I was now proposing.

> and define it from the `process_actions` functions instead of the other way around, to avoid uselessly registering a root containing `Val_unit` if not needed.

I already use Stephen's trick to make it negligible. I think it is fine as it is.

jhjourdan commented on runtime/array.c:
```diff
@@ -522,8 +523,8 @@ static value caml_array_gather(intnat num_arrays,
-       Run memprof callback for the major allocation. */
-    res = caml_check_urgent_gc_and_callbacks (res);
+       Run memprof callbacks for the major allocation. */
+    caml_process_pending_actions_with_root (res);
```
Don't forget to assign `res`:
```suggestion
    res = caml_process_pending_actions_with_root (res);
```

gadmm replied:
Well spotted, thanks! I think this would have been easier to catch after squashing the fixup commit, but of course I should not have made it in the first place.

jhjourdan commented on runtime/caml/signals.h:
```diff
@@ -30,22 +30,53 @@
-void caml_do_urgent_gc_and_callbacks (void);
-value caml_check_urgent_gc_and_callbacks (value extra_root);
-void caml_raise_in_async_callback (value exc);
+value caml_do_actions_exn (void);
```
More coherent name: `caml_do_pending_actions_exn`.

gadmm replied:
Done, thanks.

jhjourdan reviewed (changes requested):
Apart from the bug in `caml_array_gather`, I'm happy with this version.

I have a minor comment about the name `caml_do_actions`, which I don't find coherent with the other names.

Another point: forget my recent proposal to change the definition of `caml_process_pending_actions`. I had not thought enough of it, what I was thinking about does not work.

gadmm commented:
Roadmap:
- [x] Corner cases
- [x] Review of technical soundness
- [x] Review of public interface proposal
- [x] Fine-tuning
- [ ] My question about backwards-compatibility for `caml/mlvalues.h`

jhjourdan commented:
> My question about backwards-compatibility for caml/mlvalues.h

I don't think there is any issue with this.

jhjourdan commented:
The PR is ready as is. It can be merged if CI passes.

gadmm commented:
Green light on my side as well. I am very satisfied with the final state of the PR, especially with the important improvements brought via Jacques-Henri's review. Thanks!

gasche reviewed (approved):
Approved on jhjourdan's behalf.
